src/main/scala/viper/gobra/translator/encodings/maps/MapEncoding.scala

```diff
@@ -0,0 +1,361 @@
+        val (pos, info, errT) = src.vprMeta
+        val res = in.LocalVar(Names.freshName, in.IntT(Exclusive))(m.info)
+        val vRes = ctx.typeEncoding.variable(ctx)(res)
+        seqn(
+          for {
+            isCompKey <- MapEncoding.checkKeyComparability(idx)(ctx)
+            _ <- assert(isCompKey, comparabilityErrorT) // key must be comparable
+
+            vRhs <- ctx.expr.translate(rhs)(ctx)
+            vM <- ctx.expr.translate(m)(ctx)
+            vIdx <- ctx.expr.translate(idx)(ctx)
+            _ <- local(vRes)
+
+            correspondingMapM <- getCorrespondingMap(m, keys, values)(ctx)
+
+            // cannot be replaced by `getCorrespondingMap(vRes, goT(k), goT(v)`, there is no field access to the underlyingMapField
+            correspondingMapRes = vpr.DomainFuncApp(
+              func = getMapFunc,
+              args = Seq(vRes.localVar),
+              typVarMap = Map(keyParam -> goT(keys), valueParam -> goT(values)))(pos, info, errT)
+
+            inhale = vpr.Inhale(vpr.EqCmp(correspondingMapRes, vpr.MapUpdate(correspondingMapM, vIdx, vRhs)(pos, info, errT))(pos, info, errT))(pos, info, errT)
+            _ <- write(inhale)
+          } yield vpr.FieldAssign(vpr.FieldAccess(vM, underlyingMapField(ctx))(pos, info, errT), vRes.localVar)(pos, info, errT)
+        )
+    }
+  }
+
+  /**
+    * Encodes assertions.
+    * [acc(m: map[K]V, perm)] -> acc([m].underlyingMapField, [perm])
+    */
+  override def assertion(ctx: Context): in.Assertion ==> CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+
+    default(super.assertion(ctx)) {
+      case n@ in.Access(in.Accessible.ExprAccess(exp :: ctx.Map(_, _)), perm) =>
+        val (pos, info, errT) = n.vprMeta
+        for {
+          vE <- goE(exp)
+          vP <- goE(perm)
+        } yield vpr.FieldAccessPredicate(vpr.FieldAccess(vE, underlyingMapField(ctx))(pos, info, errT), vP)(pos, info, errT)
+    }
+  }
+
+  override def finalize(col: Collector): Unit = {
+    if (isUsed) {
+      col.addMember(genDomain())
+    }
+  }
+
+  private var isUsed: Boolean = false
+
+  private val keyParam = vpr.TypeVar("K")
+  private val valueParam = vpr.TypeVar("V")
+
+  /**
+    * Generates
+    *   domain GobraMap[K,V] {
+    *     function getMap(addr: Int): Map[K,V]
+    *   }
+    */
+  private def genDomain(): vpr.Domain = {
+    vpr.Domain(
+      name = domainName,
+      functions = Seq(getMapFunc),
+      axioms = Seq(),
+      typVars = Seq(keyParam, valueParam)
+    )()
+  }
+
+  private val getMapFuncName: String = "getMap"
+  private val getMapFunc: vpr.DomainFunc = vpr.DomainFunc(
+    name = getMapFuncName,
+    formalArgs = Seq(vpr.LocalVarDecl("id", vpr.Int)()),
+    typ = vpr.MapType(keyParam, valueParam),
+  )(domainName = domainName)
+
+  /**
+    * Field of the corresponding map id
+    */
+  private def underlyingMapField(ctx: Context): vpr.Field = ctx.field.field(in.IntT(Exclusive))(ctx)
```

Comment:
you should just return the field directly, `vpr.Field(someName, vpr.Int)`

Reply:
I did this to avoid introducing a new name at the viper level. Still, I can do that